### ISSUE TITLE
The csv DataProvider is triggering on startup, even with @depends

### ISSUE THREAD

**Issue by cjshowalter** [Codeception/Codeception#5549]
#### What are you trying to achieve?
I generate a csv with all my page urls in WebRecurseCest, then with tests I want to run on every page, I'd just have to use your csv data provider to then load that list. Any test that needs the list just depends on this Cest that way. (GenerateListOfPages.py generates the csv).
```php
class WebRecurseCest
{
    public function updatingUrls(AcceptanceTester $I)
    {
        $command = escapeshellcmd('python tests/_support/Helper/GenerateListOfPages.py');
        shell_exec($command);
        $I->assertTrue(file_exists(pageUrls.csv), "pageUrls.csv Not Found.");
    }
}
```
```php
class SpellcheckCest
{ 
     /**
      * @depends WebRecurseCest:updatingUrls
      * @group test-spelling
      * @dataProvider urlProvider
      */
    public function checkingSpellingOn(AcceptanceTester $I, \Codeception\Example $url)
    {
## Code here that actually checks spelling
    }
     /**
      * @depends WebRecurseCest:updatingUrls
      */
    protected function urlProvider()
    {
        if(file_exists(pageUrls.csv))
        {
            return new CsvFileIterator(pageUrls.csv);
        }
    }
}
```
#### What do you get instead?
The csv is loaded by the data provider before WebRecurseCest is ran, even with @depends. With depends attached to the dataprovider, shouldn't it have to wait for that condition before it loads the data? If there is no csv, it will try to load it instantly and fail, instead of waiting for that condition.
This seemed like the best way to me for testing every page on a website, given only a url of your website, but I could easily be missing something...
(It works with this bash script, but it seems really hacky to me):
```bash
#!/bin/bash
# First run the web crawler for the csv file:
python tests/_support/Helper/GenerateListOfPages.py

# Run codecept, and pass on whatever args that were passed to this script:
ARGS=""
while [ "$1" != "" ]; do
	ARGS="$ARGS $1"
	shift
done
codecept $ARGS
```
### Details

* Codeception version: 2.5.4
* PHP Version: 7.2.19
* Operating System: Ubuntu 18.04

```yml
# paste suite config here
```

**Comment by cjshowalter** [comment]
Some details I didn't include, in case it helps. (This was a fresh install, no plugins or anything yet).
* Installation type: Phar
* List of installed packages (composer show)
* Suite configuration:
```yaml
actor: AcceptanceTester
modules:
    enabled:
        - PhpBrowser:
            url: <website>
        - \Helper\Acceptance
        - Asserts
```
Thanks in advance for any insight you guys may have on this!

**Comment by Naktibalda** [comment]
Relevant code is in [Test\Loader\Cest::loadTests](https://github.com/Codeception/Codeception/blob/3.0.1/src/Codeception/Test/Loader/Cest.php#L56-L96), dataProvider method is executed at load time and all results are handled in the same way as `@example`s and added as separate tests to test suite.

I haven't looked at `\PHPUnit\Framework\DataProviderTestSuite` but I think that the best option would be to extend or wrap it and make it execute dataProvider method at the last possible moment.

**Comment by Naktibalda** [comment]
I think that it would be better to move your python script execution to dataProvider and execute it on the first invokation only.

**Comment by Naktibalda** [comment]
@cjshowalter have you tried to implement my suggestion?

> I think that it would be better to move your python script execution to dataProvider and execute it on the first invokation only.

It is very unlikely that data provider functionality will be modified to work the way you wanted, because it would likely break expectations of other users.

**Comment by cjshowalter** [comment]
No, I really like your suggestion, thanks! I was just out on another project for a while, sorry for that. Kinda going off of this, is there a "standard" place to launch scripts from when codecpet starts up? (before the data provider init gets called, I mean). This would allow me to just put the cvs generator there, and keep the data provider functionality the same for everyone too. If not, your first suggestion should work fine. I could also use this to generate other files at the beginning that some tests will need too.

**Comment by cjshowalter** [comment]
@Naktibalda I tried your suggestion and it works great! I'm just afraid of an update re-writing that file or something similar. If updates could do that, the bash script route doesn't seem that bad lol. There's quite a bit I want to do before codecept runs, so throwing it all into that one script should be easier to maintain than having the individual scripts called throughout codecept's init. Nothing should break with any codeception update this way.
